nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -1777,12 +1777,290 @@ impl ArrayPrototype {
+    /// > is not provided.
+    /// >
+    /// > reduceRight does not directly mutate the object on which it is called
+    /// > but the object may be mutated by the calls to callbackfn.
+    /// >
+    /// > The range of elements processed by reduceRight is set before the
+    /// > first call to callbackfn. Elements that are appended to the array
+    /// > after the call to reduceRight begins will not be visited by
+    /// > callbackfn. If existing elements of the array are changed by
+    /// > callbackfn, their value as passed to callbackfn will be the value at
+    /// > the time reduceRight visits them; elements that are deleted after the
+    /// > call to reduceRight begins and before being visited are not visited.
+    ///
+    /// > Note 2
+    /// >
+    /// > This method is intentionally generic; it does not require that its
+    /// > this value be an Array. Therefore it can be transferred to other
+    /// > kinds of objects for use as a method.
+    fn reduce_right(
+        agent: &mut Agent,
+        this_value: Value,
+        arguments: ArgumentsList,
+    ) -> JsResult<Value> {
+        let callback_fn = arguments.get(0);
+        let initial_value = if arguments.len() >= 2 {
+            Some(arguments.get(1))
+        } else {
+            None
+        };
+
+        // 1. Let O be ? ToObject(this value).
+        let o = to_object(agent, this_value)?;
+
+        // 2. Let len be ? LengthOfArrayLike(O).
+        let len = length_of_array_like(agent, o)?;
+
+        // 3. If IsCallable(callbackfn) is false, throw a TypeError exception.
+        let Some(callback_fn) = is_callable(callback_fn) else {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "Callback function is not a function",
+            ));
+        };
+
+        // 4. If len = 0 and initialValue is not present, throw a TypeError exception.
+        if len == 0 && initial_value.is_none() {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "Array length is 0 and no initial value provided",
+            ));
+        }
+
+        // 5. Let k be len - 1.
+        let mut k = len - 1;
+        // 6. Let accumulator be undefined.
+        // 7. If initialValue is present, then
+        // a. Set accumulator to initialValue.
+        let mut accumulator = initial_value.unwrap_or(Value::Undefined);
+
+        // 8. Else,
+        if initial_value.is_none() {
+            // a. Let kPresent be false.
+            let mut k_present = false;
+
+            // b. Repeat, while kPresent is false and k ≥ 0,
+            while !k_present && k >= 0 {
+                // i. Let Pk be ! ToString(𝔽(k)).
+                // NOTE: Guaranteed to be a valid u32 because it's max is the length of an array-like object.
+                let pk = PropertyKey::from(k as u32);
```

Comment:
**issue:** Ah, the u32 strikes here too. We're still actually guaranteed to get a valid PropertyKey though, since LengthOfArrayLike returns a safe integer. Just needs to be `try_from(k).unwrap()`.

Reply:
Oops